File: core/src/main/java/y2k/joyreactor/services/requests/CreateCommentRequest.java
package y2k.joyreactor.services.requests;

import rx.Observable;
import y2k.joyreactor.common.ObservableUtils;
import y2k.joyreactor.http.HttpClient;

import java.io.IOException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by y2k on 19/10/15.
 */
public class CreateCommentRequest {

    public static final Pattern TOKEN_REGEX = Pattern.compile("var token = '(.+?)'");
    private String postId;
    private String commentId;

    public CreateCommentRequest(String postId, String commentId) {
        this.postId = postId;
        this.commentId = commentId;
    }

    public Observable<Void> request(String commentText) {
        return ObservableUtils.create(() -> {
            HttpClient.getInstance()
                    .beginForm()
                    .put("parent_id", commentId == null ? "0" : commentId)
                    .put("post_id", postId)
                    .put("token", getToken())
                    .put("comment_text", commentText)
                    .putHeader("X-Requested-With", "XMLHttpRequest")
                    .putHeader("Referer", "http://joyreactor.cc/post/" + postId)
                    .send("http://joyreactor.cc/post_comment/create");
            return null;
        });
    }

    private String getToken() throws IOException {
        String document = HttpClient.getInstance().getText("http://joyreactor.cc/donate");
        Matcher m = TOKEN_REGEX.matcher(document);
        if (!m.find()) throw new IllegalStateException();
        return m.group(1);
    }
}

File: core/src/main/kotlin/y2k/joyreactor/services/requests/CreateCommentRequest.kt
package y2k.joyreactor.services.requests

import y2k.joyreactor.common.async.CompletableFuture
import y2k.joyreactor.common.http.HttpClient
import y2k.joyreactor.common.postAsync
import java.util.regex.Pattern

/**
 * Created by y2k on 19/10/15.
 */
class CreateCommentRequest(
    private val httpClient: HttpClient) : Function2<Long, String, CompletableFuture<*>> {

    private val commentId: String? = null

    override fun invoke(postId: Long, commentText: String): CompletableFuture<*> {
        return httpClient
            .buildRequest()
            .addField("parent_id", commentId ?: "0")
            .addField("post_id", "" + postId)
            .addField("token", getToken())
            .addField("comment_text", commentText)
            .putHeader("X-Requested-With", "XMLHttpRequest")
            .putHeader("Referer", "http://joyreactor.cc/post/" + postId)
            .postAsync("http://joyreactor.cc/post_comment/create")
    }

    private fun getToken(): String {
        val document = httpClient.getText("http://joyreactor.cc/donate")
        val m = TOKEN_REGEX.matcher(document)
        if (!m.find()) throw IllegalStateException()
        return m.group(1)
    }

    companion object {

        val TOKEN_REGEX = Pattern.compile("var token = '(.+?)'")
    }
}